File: StripMate/Utils/CollageBuilder.swift
import UIKit
import CoreImage
import CoreImage.CIFilterBuiltins

// MARK: - Collage Aspect Ratio

public enum CollageAspectRatio: String, CaseIterable {
    case portrait   // 9:16
    case instagram  // 4:5
    case square     // 1:1

    public var label: String {
        switch self {
        case .portrait: return "9:16"
        case .instagram: return "4:5"
        case .square: return "1:1"
        }
    }

    public var width: CGFloat { 1080 }

    public var height: CGFloat {
        switch self {
        case .portrait: return 1920
        case .instagram: return 1350
        case .square: return 1080
        }
    }

    public var ratio: CGFloat { width / height }
}

// MARK: - Collage Layout

/// Layout options for combining 2-4 photos into a single collage strip.
public enum CollageLayout: CaseIterable, Identifiable {
    // 2-photo layouts
    case twoHorizontal        // side by side
    case twoVertical          // top/bottom
    case twoDiagonal          // diagonal split — left-top big, right-bottom big
    case twoLeftWide          // 70/30 left dominant

    // 3-photo layouts
    case threeLeftLarge       // 1 big left + 2 small stacked right
    case threeTopLarge        // 1 big top + 2 small side-by-side bottom
    case threeRightLarge      // 2 small stacked left + 1 big right
    case threeBottomLarge     // 2 small top + 1 big bottom
    case threeEqualRows       // 3 horizontal rows equal
    case threeEqualCols       // 3 vertical columns equal

    // 4-photo layouts
    case fourGrid             // 2x2 grid
    case fourTopRow           // 1 big top + 3 small bottom
    case fourBottomRow        // 3 small top + 1 big bottom
    case fourLeftCol          // 1 big left + 3 stacked right
    case fourCenterFocus      // big center overlap style: top-left, top-right, bottom-left, bottom-right with center emphasis

    public var id: String {
        switch self {
        case .twoHorizontal: return "twoH"
        case .twoVertical: return "twoV"
        case .twoDiagonal: return "twoDiag"
        case .twoLeftWide: return "twoLW"
        case .threeLeftLarge: return "threeL"
        case .threeTopLarge: return "threeT"
        case .threeRightLarge: return "threeR"
        case .threeBottomLarge: return "threeB"
        case .threeEqualRows: return "threeER"
        case .threeEqualCols: return "threeEC"
        case .fourGrid: return "fourG"
        case .fourTopRow: return "fourTR"
        case .fourBottomRow: return "fourBR"
        case .fourLeftCol: return "fourLC"
        case .fourCenterFocus: return "fourCF"
        }
    }

    public var photoCount: Int {
        switch self {
        case .twoHorizontal, .twoVertical, .twoDiagonal, .twoLeftWide: return 2
        case .threeLeftLarge, .threeTopLarge, .threeRightLarge, .threeBottomLarge,
             .threeEqualRows, .threeEqualCols: return 3
        case .fourGrid, .fourTopRow, .fourBottomRow, .fourLeftCol, .fourCenterFocus: return 4
        }
    }

    /// Layouts available for a given photo count.
    public static func layouts(for count: Int) -> [CollageLayout] {
        allCases.filter { $0.photoCount == count }
    }
}

// MARK: - Photo Transform

/// Per-photo pan/zoom state for interactive collage editing.
/// offset is normalized (0-1 range relative to overflow), scale >= 1.0.
public struct PhotoTransform: Equatable {
    public var offset: CGSize = .zero   // normalized pan offset (-1...1)
    public var scale: CGFloat = 1.0     // pinch-to-zoom (1.0 = aspect fill)

    public static let identity = PhotoTransform()
}

// MARK: - Collage Background

public enum CollageBackground: String, CaseIterable {
    case black
    case white
    case blurFill
}

// MARK: - Collage Builder

/// Renders an array of UIImages into a single portrait collage image.
public enum CollageBuilder {

    /// Builds a collage from the given images using the specified layout.
    /// `transforms` allows per-photo pan/zoom adjustments from the interactive editor.
    public static func build(
        images: [UIImage],
        layout: CollageLayout,
        gap: CGFloat = 4,
        background: CollageBackground = .black,
        cornerStyle: CollageCornerStyle = .rounded,
        aspectRatio: CollageAspectRatio = .portrait,
        transforms: [Int: PhotoTransform] = [:]
    ) -> UIImage? {
        guard images.count >= layout.photoCount else { return nil }

        let normalized = images.map { $0.normalizedOrientation().resizedToMax(dimension: 1440) }

        let size = CGSize(width: aspectRatio.width, height: aspectRatio.height)
        let renderer = UIGraphicsImageRenderer(size: size)
        let radius: CGFloat = cornerStyle == .rounded ? 24 : 0

        return autoreleasepool {
            renderer.image { context in
                // Background
                switch background {
                case .black:
                    UIColor.black.setFill()
                    context.fill(CGRect(origin: .zero, size: size))

                case .white:
                    UIColor.white.setFill()
                    context.fill(CGRect(origin: .zero, size: size))

                case .blurFill:
                    drawBlurredBackground(
                        from: normalized.first,
                        in: CGRect(origin: .zero, size: size),
                        context: context.cgContext
                    )
                }

                let frames = cellFrames(for: layout, in: size, gap: gap)

                for (index, frame) in frames.enumerated() where index < normalized.count {
                    let img = normalized[index]
                    let transform = transforms[index] ?? .identity
                    drawImage(img, in: frame, cornerRadius: radius, transform: transform, context: context.cgContext)
                }
            }
        }
    }

    // MARK: - Frame Calculation

    public static func cellFrames(for layout: CollageLayout, in size: CGSize, gap: CGFloat = 4) -> [CGRect] {
        let w = size.width
        let h = size.height
        let g = gap

        switch layout {
        // ── 2-photo ────────────────────────────────────
        case .twoHorizontal:
            let cellW = (w - g) / 2
            return [
                CGRect(x: 0, y: 0, width: cellW, height: h),
                CGRect(x: cellW + g, y: 0, width: cellW, height: h)
            ]

        case .twoVertical:
            let cellH = (h - g) / 2
            return [
                CGRect(x: 0, y: 0, width: w, height: cellH),
                CGRect(x: 0, y: cellH + g, width: w, height: cellH)
            ]

        case .twoDiagonal:
            // Top-left large (60%) + bottom-right large (60%) with overlap gap
            let splitX = w * 0.55
            let splitY = h * 0.55
            return [
                CGRect(x: 0, y: 0, width: splitX - g / 2, height: splitY - g / 2),
                CGRect(x: splitX + g / 2, y: splitY + g / 2, width: w - splitX - g / 2, height: h - splitY - g / 2)
            ]

        case .twoLeftWide:
            // 70/30 left dominant
            let leftW = (w - g) * 0.7
            return [
                CGRect(x: 0, y: 0, width: leftW, height: h),
                CGRect(x: leftW + g, y: 0, width: w - leftW - g, height: h)
            ]

        // ── 3-photo ────────────────────────────────────
        case .threeLeftLarge:
            let leftW = (w - g) * 0.6
            let rightW = w - leftW - g
            let rightH = (h - g) / 2
            return [
                CGRect(x: 0, y: 0, width: leftW, height: h),
                CGRect(x: leftW + g, y: 0, width: rightW, height: rightH),
                CGRect(x: leftW + g, y: rightH + g, width: rightW, height: rightH)
            ]

        case .threeTopLarge:
            let topH = (h - g) * 0.6
            let bottomH = h - topH - g
            let cellW = (w - g) / 2
            return [
                CGRect(x: 0, y: 0, width: w, height: topH),
                CGRect(x: 0, y: topH + g, width: cellW, height: bottomH),
                CGRect(x: cellW + g, y: topH + g, width: cellW, height: bottomH)
            ]

        case .threeRightLarge:
            let rightW = (w - g) * 0.6
            let leftW = w - rightW - g
            let leftH = (h - g) / 2
            return [
                CGRect(x: 0, y: 0, width: leftW, height: leftH),
                CGRect(x: 0, y: leftH + g, width: leftW, height: leftH),
                CGRect(x: leftW + g, y: 0, width: rightW, height: h)
            ]

        case .threeBottomLarge:
            let bottomH = (h - g) * 0.6
            let topH = h - bottomH - g
            let cellW = (w - g) / 2
            return [
                CGRect(x: 0, y: 0, width: cellW, height: topH),
                CGRect(x: cellW + g, y: 0, width: cellW, height: topH),
                CGRect(x: 0, y: topH + g, width: w, height: bottomH)
            ]

        case .threeEqualRows:
            let cellH = (h - 2 * g) / 3
            return [
                CGRect(x: 0, y: 0, width: w, height: cellH),
                CGRect(x: 0, y: cellH + g, width: w, height: cellH),
                CGRect(x: 0, y: 2 * (cellH + g), width: w, height: cellH)
            ]

        case .threeEqualCols:
            let cellW = (w - 2 * g) / 3
            return [
                CGRect(x: 0, y: 0, width: cellW, height: h),
                CGRect(x: cellW + g, y: 0, width: cellW, height: h),
                CGRect(x: 2 * (cellW + g), y: 0, width: cellW, height: h)
            ]

        // ── 4-photo ────────────────────────────────────
        case .fourGrid:
            let cellW = (w - g) / 2
            let cellH = (h - g) / 2
            return [
                CGRect(x: 0, y: 0, width: cellW, height: cellH),
                CGRect(x: cellW + g, y: 0, width: cellW, height: cellH),
                CGRect(x: 0, y: cellH + g, width: cellW, height: cellH),
                CGRect(x: cellW + g, y: cellH + g, width: cellW, height: cellH)
            ]

        case .fourTopRow:
            // 1 big top (60%) + 3 small bottom
            let topH = (h - g) * 0.6
            let bottomH = h - topH - g
            let cellW = (w - 2 * g) / 3
            return [
                CGRect(x: 0, y: 0, width: w, height: topH),
                CGRect(x: 0, y: topH + g, width: cellW, height: bottomH),
                CGRect(x: cellW + g, y: topH + g, width: cellW, height: bottomH),
                CGRect(x: 2 * (cellW + g), y: topH + g, width: cellW, height: bottomH)
            ]

        case .fourBottomRow:
            // 3 small top + 1 big bottom (60%)
            let bottomH = (h - g) * 0.6
            let topH = h - bottomH - g
            let cellW = (w - 2 * g) / 3
            return [
                CGRect(x: 0, y: 0, width: cellW, height: topH),
                CGRect(x: cellW + g, y: 0, width: cellW, height: topH),
                CGRect(x: 2 * (cellW + g), y: 0, width: cellW, height: topH),
                CGRect(x: 0, y: topH + g, width: w, height: bottomH)
            ]

        case .fourLeftCol:
            // 1 big left (60%) + 3 stacked right
            let leftW = (w - g) * 0.6
            let rightW = w - leftW - g
            let cellH = (h - 2 * g) / 3
            return [
                CGRect(x: 0, y: 0, width: leftW, height: h),
                CGRect(x: leftW + g, y: 0, width: rightW, height: cellH),
                CGRect(x: leftW + g, y: cellH + g, width: rightW, height: cellH),
                CGRect(x: leftW + g, y: 2 * (cellH + g), width: rightW, height: cellH)
            ]

        case .fourCenterFocus:
            // Center photo large, 3 smaller around edges
            // Layout: top-left small, top-right small, full-width center band, bottom small
            let sideH = (h - 2 * g) * 0.25
            let centerH = (h - 2 * g) * 0.5
            let cellW = (w - g) / 2
            return [
                CGRect(x: 0, y: 0, width: cellW, height: sideH),
                CGRect(x: cellW + g, y: 0, width: cellW, height: sideH),
                CGRect(x: 0, y: sideH + g, width: w, height: centerH),
                CGRect(x: 0, y: sideH + centerH + 2 * g, width: w, height: sideH)
            ]
        }
    }

    // MARK: - Drawing

    private static func drawImage(_ image: UIImage, in rect: CGRect, cornerRadius: CGFloat, transform: PhotoTransform = .identity, context: CGContext) {
        context.saveGState()

        let path = UIBezierPath(roundedRect: rect, cornerRadius: cornerRadius)
        context.addPath(path.cgPath)
        context.clip()

        var drawRect = aspectFillRect(for: image.size, in: rect)

        // Apply user's zoom (scale up from center of cell)
        if transform.scale > 1.0 {
            let newW = drawRect.width * transform.scale
            let newH = drawRect.height * transform.scale
            drawRect = CGRect(
                x: drawRect.midX - newW / 2,
                y: drawRect.midY - newH / 2,
                width: newW,
                height: newH
            )
        }

        // Apply user's pan offset (normalized: -1...1 maps to max overflow)
        let overflowX = (drawRect.width - rect.width) / 2
        let overflowY = (drawRect.height - rect.height) / 2
        drawRect.origin.x += transform.offset.width * overflowX
        drawRect.origin.y += transform.offset.height * overflowY

        image.draw(in: drawRect)

        context.restoreGState()
    }

    /// Draws the first image as an aspect-fill blurred background.
    private static func drawBlurredBackground(
        from image: UIImage?,
        in rect: CGRect,
        context: CGContext
    ) {
        guard let image = image, let ciImage = CIImage(image: image) else {
            UIColor.black.setFill()
            UIBezierPath(rect: rect).fill()
            return
        }

        let ciContext = CIContext(options: [.useSoftwareRenderer: false])

        let blur = CIFilter.gaussianBlur()
        blur.inputImage = ciImage
        blur.radius = 30

        guard let output = blur.outputImage else {
            UIColor.black.setFill()
            UIBezierPath(rect: rect).fill()
            return
        }

        // Crop to original extent (blur expands edges)
        let cropped = output.cropped(to: ciImage.extent)

        guard let cgBlurred = ciContext.createCGImage(cropped, from: ciImage.extent) else {
            UIColor.black.setFill()
            UIBezierPath(rect: rect).fill()
            return
        }

        let blurredUIImage = UIImage(cgImage: cgBlurred)
        let fillRect = aspectFillRect(for: blurredUIImage.size, in: rect)

        context.saveGState()
        context.addPath(UIBezierPath(rect: rect).cgPath)
        context.clip()
        blurredUIImage.draw(in: fillRect)
        context.restoreGState()
    }

    /// Calculates a rect that aspect-fills `imageSize` into `bounds`.
    private static func aspectFillRect(for imageSize: CGSize, in bounds: CGRect) -> CGRect {
        let imageRatio = imageSize.width / imageSize.height
        let boundsRatio = bounds.width / bounds.height

        var drawSize: CGSize
        if imageRatio > boundsRatio {
            drawSize = CGSize(width: bounds.height * imageRatio, height: bounds.height)
        } else {
            drawSize = CGSize(width: bounds.width, height: bounds.width / imageRatio)
        }

        let x = bounds.midX - drawSize.width / 2
        let y = bounds.midY - drawSize.height / 2
        return CGRect(origin: CGPoint(x: x, y: y), size: drawSize)
    }
}

File: android/app/src/main/java/com/celalbasaran/stripmate/util/CollageBuilder.kt
package com.celalbasaran.stripmate.util

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import com.celalbasaran.stripmate.data.model.CollageAspectRatio
import com.celalbasaran.stripmate.data.model.CollageBackground
import com.celalbasaran.stripmate.data.model.CollageCornerStyle
import com.celalbasaran.stripmate.data.model.CollageLayout
import com.celalbasaran.stripmate.data.model.PhotoTransform

object CollageBuilder {

    fun build(
        images: List<Bitmap>,
        layout: CollageLayout,
        gap: Float = 4f,
        background: CollageBackground = CollageBackground.BLACK,
        cornerStyle: CollageCornerStyle = CollageCornerStyle.ROUNDED,
        aspectRatio: CollageAspectRatio = CollageAspectRatio.PORTRAIT,
        transforms: Map<Int, PhotoTransform> = emptyMap()
    ): Bitmap {
        val outputW = aspectRatio.width.toInt()
        val outputH = aspectRatio.height.toInt()
        val downscaled = images.map { downscaleToMax(it, outputH) }
        val radius = if (cornerStyle == CollageCornerStyle.ROUNDED) 24f else 0f

        val output = Bitmap.createBitmap(outputW, outputH, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(output)

        when (background) {
            CollageBackground.BLACK -> canvas.drawColor(android.graphics.Color.BLACK)
            CollageBackground.WHITE -> canvas.drawColor(android.graphics.Color.WHITE)
            CollageBackground.BLUR_FILL -> {
                if (downscaled.isNotEmpty()) {
                    val blurred = stackBlur(downscaled[0], 30)
                    val paint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG)
                    drawImageInRect(canvas, blurred, RectF(0f, 0f, outputW.toFloat(), outputH.toFloat()), paint)
                    blurred.recycle()
                } else {
                    canvas.drawColor(android.graphics.Color.BLACK)
                }
            }
        }

        val cells = getCells(layout, gap, aspectRatio)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG)

        for (i in cells.indices) {
            if (i >= downscaled.size) break
            val rect = cells[i]
            val transform = transforms[i] ?: PhotoTransform()
            drawImageInCell(canvas, downscaled[i], rect, radius, paint, transform)
        }

        return output
    }

    fun getCells(
        layout: CollageLayout,
        gap: Float = 4f,
        aspectRatio: CollageAspectRatio = CollageAspectRatio.PORTRAIT
    ): List<RectF> {
        val w = aspectRatio.width
        val h = aspectRatio.height
        val g = gap

        return when (layout) {
            // ── 2-photo ────────────────────────────────
            CollageLayout.TWO_HORIZONTAL -> {
                val cellW = (w - g) / 2f
                listOf(
                    RectF(0f, 0f, cellW, h),
                    RectF(cellW + g, 0f, w, h)
                )
            }
            CollageLayout.TWO_VERTICAL -> {
                val cellH = (h - g) / 2f
                listOf(
                    RectF(0f, 0f, w, cellH),
                    RectF(0f, cellH + g, w, h)
                )
            }
            CollageLayout.TWO_DIAGONAL -> {
                val splitX = w * 0.55f
                val splitY = h * 0.55f
                listOf(
                    RectF(0f, 0f, splitX - g / 2f, splitY - g / 2f),
                    RectF(splitX + g / 2f, splitY + g / 2f, w, h)
                )
            }
            CollageLayout.TWO_LEFT_WIDE -> {
                val leftW = (w - g) * 0.7f
                listOf(
                    RectF(0f, 0f, leftW, h),
                    RectF(leftW + g, 0f, w, h)
                )
            }

            // ── 3-photo ────────────────────────────────
            CollageLayout.THREE_LEFT_LARGE -> {
                val leftW = (w - g) * 0.6f
                val rightH = (h - g) / 2f
                listOf(
                    RectF(0f, 0f, leftW, h),
                    RectF(leftW + g, 0f, w, rightH),
                    RectF(leftW + g, rightH + g, w, h)
                )
            }
            CollageLayout.THREE_TOP_LARGE -> {
                val topH = (h - g) * 0.6f
                val cellW = (w - g) / 2f
                listOf(
                    RectF(0f, 0f, w, topH),
                    RectF(0f, topH + g, cellW, h),
                    RectF(cellW + g, topH + g, w, h)
                )
            }
            CollageLayout.THREE_RIGHT_LARGE -> {
                val rightW = (w - g) * 0.6f
                val leftW = w - rightW - g
                val leftH = (h - g) / 2f
                listOf(
                    RectF(0f, 0f, leftW, leftH),
                    RectF(0f, leftH + g, leftW, h),
                    RectF(leftW + g, 0f, w, h)
                )
            }
            CollageLayout.THREE_BOTTOM_LARGE -> {
                val bottomH = (h - g) * 0.6f
                val topH = h - bottomH - g
                val cellW = (w - g) / 2f
                listOf(
                    RectF(0f, 0f, cellW, topH),
                    RectF(cellW + g, 0f, w, topH),
                    RectF(0f, topH + g, w, h)
                )
            }
            CollageLayout.THREE_EQUAL_ROWS -> {
                val cellH = (h - 2f * g) / 3f
                listOf(
                    RectF(0f, 0f, w, cellH),
                    RectF(0f, cellH + g, w, 2f * cellH + g),
                    RectF(0f, 2f * (cellH + g), w, h)
                )
            }
            CollageLayout.THREE_EQUAL_COLS -> {
                val cellW = (w - 2f * g) / 3f
                listOf(
                    RectF(0f, 0f, cellW, h),
                    RectF(cellW + g, 0f, 2f * cellW + g, h),
                    RectF(2f * (cellW + g), 0f, w, h)
                )
            }

            // ── 4-photo ────────────────────────────────
            CollageLayout.FOUR_GRID -> {
                val cellW = (w - g) / 2f
                val cellH = (h - g) / 2f
                listOf(
                    RectF(0f, 0f, cellW, cellH),
                    RectF(cellW + g, 0f, w, cellH),
                    RectF(0f, cellH + g, cellW, h),
                    RectF(cellW + g, cellH + g, w, h)
                )
            }
            CollageLayout.FOUR_TOP_ROW -> {
                val topH = (h - g) * 0.6f
                val bottomH = h - topH - g
                val cellW = (w - 2f * g) / 3f
                listOf(
                    RectF(0f, 0f, w, topH),
                    RectF(0f, topH + g, cellW, h),
                    RectF(cellW + g, topH + g, 2f * cellW + g, h),
                    RectF(2f * (cellW + g), topH + g, w, h)
                )
            }
            CollageLayout.FOUR_BOTTOM_ROW -> {
                val bottomH = (h - g) * 0.6f
                val topH = h - bottomH - g
                val cellW = (w - 2f * g) / 3f
                listOf(
                    RectF(0f, 0f, cellW, topH),
                    RectF(cellW + g, 0f, 2f * cellW + g, topH),
                    RectF(2f * (cellW + g), 0f, w, topH),
                    RectF(0f, topH + g, w, h)
                )
            }
            CollageLayout.FOUR_LEFT_COL -> {
                val leftW = (w - g) * 0.6f
                val rightW = w - leftW - g
                val cellH = (h - 2f * g) / 3f
                listOf(
                    RectF(0f, 0f, leftW, h),
                    RectF(leftW + g, 0f, w, cellH),
                    RectF(leftW + g, cellH + g, w, 2f * cellH + g),
                    RectF(leftW + g, 2f * (cellH + g), w, h)
                )
            }
            CollageLayout.FOUR_CENTER_FOCUS -> {
                val sideH = (h - 2f * g) * 0.25f
                val centerH = (h - 2f * g) * 0.5f
                val cellW = (w - g) / 2f
                listOf(
                    RectF(0f, 0f, cellW, sideH),
                    RectF(cellW + g, 0f, w, sideH),
                    RectF(0f, sideH + g, w, sideH + g + centerH),
                    RectF(0f, sideH + centerH + 2f * g, w, h)
                )
            }
        }
    }

    private fun drawImageInCell(canvas: Canvas, bitmap: Bitmap, rect: RectF, cornerRadius: Float, paint: Paint, transform: PhotoTransform = PhotoTransform()) {
        canvas.save()
        val path = Path()
        if (cornerRadius > 0f) {
            path.addRoundRect(rect, cornerRadius, cornerRadius, Path.Direction.CW)
        } else {
            path.addRect(rect, Path.Direction.CW)
        }
        canvas.clipPath(path)
        drawImageInRect(canvas, bitmap, rect, paint, transform)
        canvas.restore()
    }

    /** Draws bitmap aspect-fill into the given rect, applying pan/zoom transform. */
    private fun drawImageInRect(canvas: Canvas, bitmap: Bitmap, rect: RectF, paint: Paint, transform: PhotoTransform = PhotoTransform()) {
        val cellWidth = rect.width()
        val cellHeight = rect.height()

        val bmpRatio = bitmap.width.toFloat() / bitmap.height.toFloat()
        val cellRatio = cellWidth / cellHeight

        // Calculate aspect-fill draw rect
        var drawW: Float
        var drawH: Float
        if (bmpRatio > cellRatio) {
            drawH = cellHeight
            drawW = drawH * bmpRatio
        } else {
            drawW = cellWidth
            drawH = drawW / bmpRatio
        }

        // Apply user's zoom
        if (transform.scale > 1f) {
            drawW *= transform.scale
            drawH *= transform.scale
        }

        var drawLeft = rect.centerX() - drawW / 2f
        var drawTop = rect.centerY() - drawH / 2f

        // Apply user's pan offset (normalized)
        val overflowX = (drawW - cellWidth) / 2f
        val overflowY = (drawH - cellHeight) / 2f
        drawLeft += transform.offsetX * overflowX
        drawTop += transform.offsetY * overflowY

        val dstRect = android.graphics.Rect(
            drawLeft.toInt(),
            drawTop.toInt(),
            (drawLeft + drawW).toInt(),
            (drawTop + drawH).toInt()
        )
        val srcRect = android.graphics.Rect(0, 0, bitmap.width, bitmap.height)
        canvas.drawBitmap(bitmap, srcRect, dstRect, paint)
    }

    /** Downscale a bitmap so its largest dimension is at most maxDim. */
    private fun downscaleToMax(bitmap: Bitmap, maxDim: Int): Bitmap {
        val maxSide = maxOf(bitmap.width, bitmap.height)
        if (maxSide <= maxDim) return bitmap

        val scale = maxDim.toFloat() / maxSide.toFloat()
        val newW = (bitmap.width * scale).toInt()
        val newH = (bitmap.height * scale).toInt()
        return Bitmap.createScaledBitmap(bitmap, newW, newH, true)
    }

    /**
     * Simple stack blur implementation for BLUR_FILL background.
     */
    fun stackBlur(source: Bitmap, radius: Int): Bitmap {
        if (radius < 1) return source.copy(Bitmap.Config.ARGB_8888, true)

        val scale = 0.25f
        val smallW = (source.width * scale).toInt().coerceAtLeast(1)
        val smallH = (source.height * scale).toInt().coerceAtLeast(1)
        val small = Bitmap.createScaledBitmap(source, smallW, smallH, true)

        val w = small.width
        val h = small.height
        val pixels = IntArray(w * h)
        small.getPixels(pixels, 0, w, 0, 0, w, h)

        val r = radius.coerceAtMost(w / 2).coerceAtMost(h / 2).coerceAtLeast(1)

        // Horizontal pass
        val temp = IntArray(w * h)
        for (y in 0 until h) {
            var rSum = 0L; var gSum = 0L; var bSum = 0L
            val rowStart = y * w
            for (x in -r..r) {
                val clampedX = x.coerceIn(0, w - 1)
                val pixel = pixels[rowStart + clampedX]
                rSum += (pixel shr 16) and 0xFF
                gSum += (pixel shr 8) and 0xFF
                bSum += pixel and 0xFF
            }
            val windowSize = (2 * r + 1)
            for (x in 0 until w) {
                temp[rowStart + x] = (0xFF shl 24) or
                        (((rSum / windowSize).toInt().coerceIn(0, 255)) shl 16) or
                        (((gSum / windowSize).toInt().coerceIn(0, 255)) shl 8) or
                        ((bSum / windowSize).toInt().coerceIn(0, 255))
                val addX = (x + r + 1).coerceAtMost(w - 1)
                val removeX = (x - r).coerceAtLeast(0)
                val addPixel = pixels[rowStart + addX]
                val removePixel = pixels[rowStart + removeX]
                rSum += ((addPixel shr 16) and 0xFF) - ((removePixel shr 16) and 0xFF)
                gSum += ((addPixel shr 8) and 0xFF) - ((removePixel shr 8) and 0xFF)
                bSum += (addPixel and 0xFF) - (removePixel and 0xFF)
            }
        }

        // Vertical pass
        val result = IntArray(w * h)
        for (x in 0 until w) {
            var rSum = 0L; var gSum = 0L; var bSum = 0L
            for (y in -r..r) {
                val clampedY = y.coerceIn(0, h - 1)
                val pixel = temp[clampedY * w + x]
                rSum += (pixel shr 16) and 0xFF
                gSum += (pixel shr 8) and 0xFF
                bSum += pixel and 0xFF
            }
            val windowSize = (2 * r + 1)
            for (y in 0 until h) {
                result[y * w + x] = (0xFF shl 24) or
                        (((rSum / windowSize).toInt().coerceIn(0, 255)) shl 16) or
                        (((gSum / windowSize).toInt().coerceIn(0, 255)) shl 8) or
                        ((bSum / windowSize).toInt().coerceIn(0, 255))
                val addY = (y + r + 1).coerceAtMost(h - 1)
                val removeY = (y - r).coerceAtLeast(0)
                val addPixel = temp[addY * w + x]
                val removePixel = temp[removeY * w + x]
                rSum += ((addPixel shr 16) and 0xFF) - ((removePixel shr 16) and 0xFF)
                gSum += ((addPixel shr 8) and 0xFF) - ((removePixel shr 8) and 0xFF)
                bSum += (addPixel and 0xFF) - (removePixel and 0xFF)
            }
        }

        small.setPixels(result, 0, w, 0, 0, w, h)

        val output = Bitmap.createScaledBitmap(small, source.width, source.height, true)
        if (small !== output) small.recycle()
        return output
    }
}
